### ISSUE TITLE
Instrument for interoperability

### ISSUE THREAD

**Issue by samreid** [phetsims/build-an-atom#107]
From https://github.com/phetsims/together/issues/151

I'm planning to begin on this soon.

**Comment by samreid** [comment]
An issue with BAAGameModel, sometimes the stateProperty is a string, sometimes it is a value returned by ProblemSetFactory._createProblem.  @jbphet should we transform this into a single type Property?

**Comment by samreid** [comment]
I'll need to fix https://github.com/phetsims/together/issues/214 before I can proceed here.

**Comment by jbphet** [comment]
Yes, it would be good to transform this stateProperty into a single type.  BTW, this was the first JS sim I did, and it was early in PhET's HTML5 evolution, we are likely to encounter a number of issues like this as we work on this sim.  It would be good to invest some time in evolving the code towards our current standards and conventions as we go.

**Comment by samreid** [comment]
The above commit is:

> Added tandems for game level 1 periodic table,

I tested manually and with the fuzz test, no issues discovered.

**Comment by samreid** [comment]
Things have been going smoothly.  Things that would still be good to do:
1. Make sure tandem.removeInstance is called and that items are deregistered in the instance-proxy test.
2. Try a wrapper that customizes a pooled game component, such as changing the color/position/visibility of the "check answer" button.
3. Some common items have been instrumented in the sim code but not outfitted in the -api.js file.  This will error out on startup when launched with ?brand=phet-io.  We should make sure the sim can launch and pass fuzz test for screen 3 ok.

**Comment by samreid** [comment]
I asked @jbphet for help disposing the past game views (essential for (1) above) and we are going to try to connect later on.

**Comment by samreid** [comment]
I investigated (2) by adding instance proxy support for buttons, and it seems to be working well.

**Comment by samreid** [comment]
I fixed (3) in prior commits.

**Comment by samreid** [comment]
At yesterday's meting, @ariel-phet suggested: 

> @samreid will finish instrumenting the game so it is not in a halfway state.  Shared stuff will be fixed.  Don’t fully instrument and bring to production since it is not using shred.

**Comment by samreid** [comment]
Instrumented game types in the above commit.

**Comment by samreid** [comment]
Level 1 is pretty well instrumented for the data stream.  My next steps:
1. Fix the random number generator for input stream playback
2. Make sure it is easy to get a game summary
3. Make sure it is possible to load it up in a specific game or problem type

**Comment by samreid** [comment]
My current recommendation:

Investigate instrumenting Random instances with tandem (as other instances are instrumented), and see if we can send the seed values in such a way that we can still support mirroring for testing.

**Comment by samreid** [comment]
> 2. Make sure it is easy to get a game summary

Most (all?) of the important information is passed to LevelCompletedNode, which serves as a central point (across many games) for knowing when a level is complete.  However, it feels somewhat hack-ish to put hooks into that constructor to transmit results of the game.  On the other hand, putting the message elsewhere will require code duplication and increased places where things can get out of sync.

I'll experiment with putting something in LevelCompletedNode, but we may want to consider another level of abstraction here.

**Comment by samreid** [comment]
I added the game summary in the above commits, it will look like this in the data stream:

```js
{
  "messageIndex": 14,
  "eventType": "user",
  "togetherID": "buildAnAtom.gameScreen.view.problemView_0.countsToElementProblemView.nextButton",
  "componentType": "TButton",
  "event": "fired",
  "time": 1452835383700,
  "children": [
    {
      "messageIndex": 15,
      "eventType": "model",
      "togetherID": "buildAnAtom.gameScreen.model",
      "componentType": "TBAAGameModel",
      "event": "levelCompleted",
      "time": 1452835383717,
      "parameters": {
        "level": 0,
        "maxPoints": 2,
        "problems": 1,
        "timerEnabled": true,
        "elapsedTime": 7,
        "bestTimes": 7,
        "newBestTime": false
      }
    }
  ]
}
```

**Comment by samreid** [comment]
After the above commit, any of the four levels can be launched immediately with query parameters like:

> together.expressions=[["buildAnAtom.gameScreen.model","startGameLevel",["periodic-table-game"]]]&screens=3

**Comment by samreid** [comment]
Forthcoming changes allow the game problem types to be selected, for instance:

> together.expressions=[["buildAnAtom.gameScreen.model","setAllowedProblemTypesByLevel",[[["counts-to-charge"],[],[],[]]]],["buildAnAtom.gameScreen.model","startGameLevel",["periodic-table-game"]]]&screens=3

Breaking this down, we have the first command which chooses the type of problem for level 1
> "buildAnAtom.gameScreen.model","setAllowedProblemTypesByLevel",[[["counts-to-charge"],[],[],[]]]

and the second command which starts level 1 with the selected problem type:
> "buildAnAtom.gameScreen.model","startGameLevel",["periodic-table-game"]]

The names aren't great, but this works.  This will be a much larger investment if we want to iron out all of the names.

**Comment by samreid** [comment]
We could do something similar to restrict the number of problems per game, but that would be a more intensive refactoring in the BAA code.

**Comment by samreid** [comment]
To end a game after a single challenge, you could listen for events of this form:

```js
{
  "messageIndex": 5,
  "eventType": "user",
  "togetherID": "buildAnAtom.gameScreen.view.problemView_0.countsToChargeProblemView.nextButton",
  "componentType": "TButton",
  "event": "fired",
  "time": 1452842520333,
  "children":[/*...*/]
}
```

and then close the iframe.  This seems straightforward and hence does not need to be implemented as a feasibility test.  We could choose to implement it as a demo though.

**Comment by samreid** [comment]
I've explored selection of level/challenges and it is working well.  I'll leave https://github.com/phetsims/together/issues/215 for further streamline/demo possibilities, but as far as this issue is concerned that part is complete.  The last thing to do here is grind through all of the problem types and instrument their components.  We also need to instrument the vibe "continue" button at the end of a game.

**Comment by samreid** [comment]
I got through 14 of the 15 problem types, will commit shortly.  Still need to hit `symbol-to-schematic`, will be extra work since you can drag electrons, protons, neutrons into the atom.

**Comment by samreid** [comment]
I've created new issues for the remaining work for the game instrumentation.  We discussed that after I complete the game instrumentation, I would leave instrumentation of screens 1-2 for someone else to instrument in the future.  Unassigning.

**Comment by samreid** [comment]
A reminder from today's status meeting that @aadish will be working on this code as I continue instrumentation, and I'll need to synchronize with him.

**Comment by samreid** [comment]
At today's meeting, we concluded that when @jbphet begins on this (hopefully within a week or so), @samreid will be on call for providing assistance.

**Comment by samreid** [comment]
As discussed in #137 

The instrumentation from https://github.com/phetsims/build-an-atom/issues/107#issuecomment-253622172 is more extensive than I expected.  The instance proxies wrapper reveals the following components in screen 1 are already instrumented:

* the particle positions
* the particle input listeners
* Show Element Name Check Box
* Show Neutral Or Ion Check Box
* Show Stable Or Unstable Check Box
* Label Visualization Control Panel
* Orbits Radio Button
* Cloud Radio Button
* Reset All Button

So the 1st screen sim is about 20% instrumented already.

You may wish to use ?screens=1 as you instrument to restrict error messages and simplify the process.  Once ?screens=1 is complete, you can move to ?screens=2, etc.

**Comment by samreid** [comment]
We are outfitting this for phet-io in part for usage in a 3rd party's research for accessibility, such as being able to throw electrons with gestures.

**Comment by samreid** [comment]
As noted in https://github.com/phetsims/phet-io/issues/730, there is partial instrumentation in the game screen which will need to be updated as we proceed.

**Comment by samreid** [comment]
It's working great:

![image](https://cloud.githubusercontent.com/assets/679486/21036661/d0e08490-bd84-11e6-89b8-7d755017d349.png)

Unassigned for now, I'll work on this in conjunction with @jbphet.

**Comment by jbphet** [comment]
Work on this was deferred for a period of time while @samreid and I worked on instrumenting Balloons and Static Electricity.  For that sim, we worked through a checklist to prompt us for each step of the instrumentation process.  I found this to be quite helpful and, even though it's a little late in the game, I'm going to add that same checklist here.  This will help me to get back up to speed on exactly where we are on this sim and to keep our efforts organized as we move forward.

The checklist shown below is being used to guide instrumentation of the **first two screens only**.  A separate checklist will be used for the game screen, which we plan to do after completing the first two.

In order to support https://github.com/phetsims/a11y-research/issues/2 and to have another sim instrumented for phet-io.

 - [x] [Understand the goal](#understand-the-goal)
 - [x] [Code review](#code-review)
 - [x] [Import Tandem to main.js](#import-tandem-to-main.js)
 - [x] [Create root tandem](#create-root-tandem)
 - [x] [Pass tandems to sim screens](#pass-tandems-to-sim-screens)
 - [x] [Pass tandems to model and view](#pass-tandems-to-model-and-view)
 - [ ] [Pass tandems to all constructors](#pass-tandems-to-all-constructors)
   + See also [Testing while adding Tandem instances](#testing-while-adding-tandem-instances). Test with `?phet-io.validateTandems=false` since common code components don't yet have tandems.
 - [ ] [Pass tandem instances to common code components](#pass-tandem-instances-to-common-code-components)
   + See also [Testing while adding Tandem instances](#testing-while-adding-tandem-instances).
 - [ ] [Replace types with tandemized equivalents](#replace-types-with-tandemized-equivalents)
 - [ ] [Instrument any Uninstrumented Common Code Components Used by the Sim](#instrument-any-uninstrumented-common-code-components-used-by-the-sim)
 - [ ] [Instrument new component types defined by the sim](#instrument-new-component-types-defined-by-the-sim)
 - [ ] [Add simulation specific support](#add-simulation-specific-support)
 - [ ] Are random numbers using `phet.joist.random`, and all doing so after modules are declared (non-statically)?  For
 instance, the following methods (and perhaps others) should not be used: `Math.random`, `_.shuffle`, `_.sample`, `_.random`
 - [ ] `undefined` values cannot be saved from the phet-io state save, sims should be written to use `null` instead.
 - [ ] Units in TNumber should be abbreviated, see https://github.com/phetsims/phet-io/issues/530
 - [ ] Update these instructions if you find them to be incomplete, inconsistent or incorrect.

**Comment by jbphet** [comment]
For next time, I should continue with the item 'Pass tandem instances to common code components' starting from line 200 of `AtomView.js`.

**Comment by jbphet** [comment]
I believe I've completed the step labelled "Pass tandem instances to common code components".  The next step in the checklist is "Replace types with tandemized equivalents", but things have recently changed and there are only a handful of component types that need this any more, and I believe they are all covered too.  So, next time I should start on the step "Instrument any Uninstrumented Common Code Components Used by the Sim" and I need to remember that shred is a common code repository and there is a bunch of things used in the sim that probably need to be instrumented.  Also, I noticed today that AccordionBox needs instrumenting, and @samreid may or may not have gotten to it by the next time I work on this.

**Comment by jbphet** [comment]
Note that as of now the state wrapper mostly works, but the atom view, including the element label, ion-vs-neutral indicator, and the red x marker don't update when particles are moved in and out.  I think this is because the files in shred haven't been fully instrumented, so once they are, I should test this capability right away and work out any problems.

**Comment by samreid** [comment]
I'm available to start assisting on the PhET-iO instrumentation for this sim.  I'll review the status and see how I can help out.

**Comment by samreid** [comment]
I confirmed in build-an-atom and shred that random numbers are using joist and there are no random lodash methods being used.

**Comment by samreid** [comment]
I'll take a look at the `Replace types with tandemized equivalents` step next.

**Comment by samreid** [comment]
Fixed above, I'll add a checkbox in the checklist for @jbphet to review my changes.

**Comment by samreid** [comment]
I didn't see any TNumbers that need units, so I checked off the "TNumber units" checkbox.

**Comment by samreid** [comment]
Turning up the priority so it will show up on the top of my todo lists.

**Comment by samreid** [comment]
Spot testing the instance proxies and data stream for the game screen seems like it is done or close to being done.  Would be good to review with @jbphet.

@jbphet let's touch base tomorrow and see the best way to continue.

**Comment by samreid** [comment]
@jbphet has taken the lead, and I'm taking a consulting role.  Unassigning myself.

**Comment by zepumph** [comment]
Hey @jbphet, I have been seeing this error in phet-io aqua tests for a few days now, and I think this is the place it stems from. I saw that you instrumented AccordionBox, now we are getting this error in BAA in this test: http://localhost/aqua/test-server/test-sims.html?ea&audioVolume=0&testBuilt=false&testDuration=10000&fuzzMouse&brand=phet-io&phetioStandalone&testSims=build-an-atom

```
Uncaught Error: Assertion failed: Tandem was required but not supplied
Error: Assertion failed: Tandem was required but not supplied
    at window.assertions.assertFunction (http://localhost/assert/js/assert.js:21:13)
    at Tandem.addInstance (http://localhost/tandem/js/Tandem.js?bust=1488930355563:80:21)
    at AccordionBox.setTandem (http://localhost/scenery/js/nodes/Node.js?bust=1488930355563:3467:22)
    at AccordionBox.set tandem [as tandem] (http://localhost/scenery/js/nodes/Node.js?bust=1488930355563:3472:32)
    at http://localhost/scenery/js/nodes/Node.js?bust=1488930355563:4763:25
    at u (http://localhost/sherpa/lib/lodash-4.17.4.min.js?bust=1488930355563:5:432)
    at Function.ru (http://localhost/sherpa/lib/lodash-4.17.4.min.js?bust=1488930355563:67:656)
    at AccordionBox.mutate (http://localhost/scenery/js/nodes/Node.js?bust=1488930355563:4750:9)
    at new AccordionBox (http://localhost/sun/js/AccordionBox.js?bust=1488930355563:299:10)
    at new BuildAnAtomView (http://localhost/build-an-atom/js/buildanatom/view/BuildAnAtomView.js?bust=1488930355563:95:25)
```

Perhaps you already know, in which case I'm sorry to bother.

**Comment by samreid** [comment]
I noticed that error too, perhaps @jbphet has made changes in the RC branch that need to be merged back to master?

**Comment by jbphet** [comment]
Indeed, the error mentioned by @zepumph is due to a commit that has not yet been propagated from the 1.4 branch to master.  I will do so within the next few days (i.e. on my next phetio day), but I want to move all such commits at the same time to avoid any confusion.

**Comment by zepumph** [comment]
Hey, I hope I didn't just make things difficult, but I just added the tandem in the above commit. It was an error that was blocking tests on the sim for other changes I am making in common code. Let me know if it is a pain and I can revert it before you merge the rest of them to master.

**Comment by jbphet** [comment]
We finished this effort for the first two screens and released the sim, and are tracking instrumentation of the game under a separate issues, see #156.  Closing.
